### ISSUE TITLE
Steam.Common - Steam ID API

### ISSUE THREAD

**Issue by ObsidianMinor** [steamframework/Steam.Framework#1]
A [Steam ID](https://developer.valvesoftware.com/wiki/SteamID) is a globally unique identifier for Steam accounts, groups, lobbies, and chat rooms.

Steam IDs are used throughout Steam, on the web and in the client. They have three formats:

 * Raw bits: 64 bits in length, sometimes called the "community ID"
 * Steam2: a common format for public user IDs
 * Steam3: a format used in debugging and games as they can support a larger variety of account types

The `SteamId` type is a read-only struct that represents the data stored in these formats. It can handle conversions from a raw values, Steam2 format strings, and Steam3 format string. It also allows retrieval of the individual pieces of information a Steam ID makes up, including account universe, type, ID, and instance.

Three other enums will need to be added to support `SteamId`: `Universe`, `AccountType`, and `Instance`.

### Details

Constructors for `SteamId` check that the ID is valid, so all `SteamId` values must be valid. Unchecked constructors may be added in the future.

`SteamId` has a number of predefined values defined in the public client header as part of the Steamworks SDK. These predefined values are listed in the API outline below. These predefined values also contain `Zero` for the default `SteamId` represented in raw as just zero.

Properties are exposed for getting the universe, type, ID, and instance of the account this `SteamId` is for.

Two constructors are provided. The first constructor is for individually setting each property of the `SteamId` (instance is range checked as it's 20 bits long and throws an out of range exception when invalid). The second constructor is for creating a individual user `SteamId` with the specified instance flags (instance is also range checked and throws an out of range exception when invalid).

Two methods are borrowed from the public client header that creates anonymous user and anonymous game server IDs. Check the Steamworks SDK for info about implementing these methods.

Three methods are defined for converting from the three `ID` formats. `FromUInt64`, `FromSteam2`, and `FromSteam3` convert from their respective formats as necessary. Invalid inputs throw `ArgumentException`.

The `Is` properties after are validity checks and simple properties pulled from public client headers. Implementations for these can be ported from there.

`ToChat`, `ToClan`, `ClanToChat`, and `ChatToClan` are also ported from client headers.

`ToStaticAccountKey` is also ported from client headers. The other `To` methods are counterparts to their `From` methods, throwing `InvalidOperationException` on incompatible IDs instead of `ArgumentException`.

`SteamId` overrides all object methods and implements `IEquatable<SteamId>`. `ToString` returns the ID in Steam3 format. It also includes a `StaticEquals` method for comparing the static properties of the IDs. Implicit and explicit operators are provided for converting to and from raw bits.

### API

```cs
public readonly struct SteamId : IEquatable<SteamId>
{
    static readonly SteamId Zero;

    // predefined from sdk
    static readonly SteamId OutOfDateGameServer;
    static readonly SteamId LanModeGameServer;
    static readonly SteamId NetYetInitializedGameServer;
    static readonly SteamId NoSteamGameServer;

    readonly ulong Value;

    Universe AccountUniverse { get; }
    AccountType AccountType { get; }
    uint AccountId { get; }
    uint AccountInstance { get; }

    SteamId(ulong id);
    SteamId(uint id, AccountType type, Universe universe, uint instance);

    static SteamId CreateAnonymousUser(Universe universe);
    static SteamId CreateAnonymousGameServer(Universe universe);
    static SteamId CreateIndividualUser(uint id, Universe universe, Instance instance);

    static SteamId FromSteam2(string id);
    static SteamId FromSteam3(string id);

    static bool IsValid(SteamId id);
    bool IsBlankAnonymousAccount { get; }
    bool IsPersistentGameServer { get; }
    bool IsGameServer { get; }
    bool IsContentServer { get; }
    bool IsClan { get; }
    bool IsChat { get; }
    bool IsIndividualAccount { get; }
    bool IsConsoleUser { get; }
    bool IsAnonymousGameServer { get; }
    bool IsAnonymousUser { get; }
    bool IsAnonymousAccount { get; }
    bool IsLobby { get; }
    
    static SteamId ToChat(SteamId id);
    static SteamId ToClan(SteamId id);
    static SteamId ChatToClan(SteamId id);
    static SteamId ClanToChat(SteamId id);

    static ulong ToStaticAccountKey(SteamId id);
    static string ToSteam2(SteamId id);
    static string ToSteam3(SteamId id);

    bool StaticEquals(SteamId other);
    bool Equals(SteamId id);
    override bool Equals(object obj);
    override bool GetHashCode();
    override string ToString();
    static implicit operator ulong(SteamId id);
    static implicit operator SteamId(ulong id);
    static bool operator ==(SteamId a, SteamId b);
    static bool operator !=(SteamId a, SteamId b);
}

public enum Universe
{
    Invalid,
    Public,
    Beta,
    Internal,
    Dev
}

public enum AccountType
{
    Invalid,
    Individual,
    Multiseat,
    GameServer,
    AnonGameServer,
    Pending,
    ContentServer,
    Clan,
    Chat,
    ConsoleUser,
    AnonUser
}

[Flags]
public enum Instance
{
    All = 0,
    Desktop = 1,
    Console = 2,
    Web = 4
}
```

# Changes

* ID was changed to Id by popular vote.
* Many methods were made static as to not look like they change state.
* Pending and Unknown tracking was removed.
* ~~What constitutes valid input has been more constrained.~~
* Moved individual account constructor to static helper method.
* Revert decision on constrained valid input.
* Move FromUInt64 to constructor. 
* Remove ToUInt64.
* Remove IsGroupChat since it was removed from public headers
* Expose underlying value

**Comment by RyadaProductions** [comment]
From my pov it doesnt make sense making this a stuct compared to a class. There are a few reasons for this.

- It implements an interface, once somebody (us, the user, the framework) cast it to the interface type it will become a referencetype allocated on the heap, this is one of my lesser concerns.

The biggest concern is that it feels like mosr of this functionality is basically screaming for a class. Since we have 6 references of SteamID living for an undefined amount of time..

- According to our talk in discord you also wanted to make most of the bool propertues have an expression body. Imo would that defeat the purpose of a struct as you will hide an if check behind a property, the property suggest it returns a value while what the intention is here is to check if the current struct is of type amd then return the result, adding (albeit minor) overhead.

In short i feel like we are bound to run into unintended behaviour with the current suggestion for the api.

**Comment by ObsidianMinor** [comment]
The reason that `SteamID` is a struct is because it's a *value*. It's represented as a ulong value in Steam, it's transferred as a ulong value 99% of the time (sometimes they send account IDs around). Ultimately making everything properties leaves the type open to alternative implementation methods. 

For example instead of having 4 fields with each variable we could do a Steam and make it set the bits directly. And instead of two bools we could have another ulong with bits set for each `Is` property.
